test/test_connectionpool.py

```diff
@@ -308,7 +311,91 @@ def kaboom(*args, **kwargs):
 
         new_pool_size = c.pool.qsize()
         self.assertEqual(initial_pool_size, new_pool_size)
-
+    
+    def test_preserves_request_header_order(self):
+        """
+        Ensure that headers provided to ConnectionPool.urlopen()
+        are sent to the underlying httplib in the same order.
+        """
+        class AbortRequest(BaseException):
+            pass
+        
+        expected_request_headers = OrderedDict([('X-Header-%d' % i, str(i)) for i in range(16)])
```

Comment:
Right now it gets inserted in alphabetical order, no? How about we just random.shuffle(...) it and check that we get it back in the same shuffled order? That should be fine with a small number of entries.